File: src/main/java/org/lanternpowered/server/network/entity/vanilla/ZombieVillagerEntityProtocol.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.entity.vanilla;

import org.lanternpowered.server.data.key.LanternKeys;
import org.lanternpowered.server.data.type.LanternProfession;
import org.lanternpowered.server.entity.LanternEntity;
import org.lanternpowered.server.network.entity.parameter.ParameterList;
import org.spongepowered.api.data.Keys;
import org.spongepowered.api.data.type.Career;
import org.spongepowered.api.data.type.Professions;

public class ZombieVillagerEntityProtocol<E extends LanternEntity> extends AbstractZombieEntityProtocol<E> {

    private int lastProfession;
    private boolean lastIsConverting;

    public ZombieVillagerEntityProtocol(E entity) {
        super(entity);
    }

    private int getProfessionId() {
        return ((LanternProfession) this.entity.get(Keys.CAREER).map(Career::getProfession)
                .orElse(Professions.FARMER)).getInternalId();
    }

    @Override
    protected String getMobType() {
        return "minecraft:zombie_villager";
    }

    @Override
    protected void spawn(ParameterList parameterList) {
        super.spawn(parameterList);
        parameterList.add(EntityParameters.ZombieVillager.PROFESSION, getProfessionId());
        parameterList.add(EntityParameters.ZombieVillager.IS_CONVERTING, this.entity.get(LanternKeys.IS_CONVERTING).orElse(false));
    }

    @Override
    protected void update(ParameterList parameterList) {
        super.update(parameterList);
        final int profession = getProfessionId();
        if (profession != this.lastProfession) {
            parameterList.add(EntityParameters.ZombieVillager.PROFESSION, profession);
            this.lastProfession = profession;
        }
        final boolean isConverting = this.entity.get(LanternKeys.IS_CONVERTING).orElse(false);
        if (isConverting != this.lastIsConverting) {
            parameterList.add(EntityParameters.ZombieVillager.IS_CONVERTING, isConverting);
            this.lastIsConverting = isConverting;
        }
    }
}

File: src/main/kotlin/org/lanternpowered/server/network/entity/vanilla/ZombieVillagerEntityProtocol.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.entity.vanilla

import org.lanternpowered.api.key.NamespacedKey
import org.lanternpowered.api.key.minecraftKey
import org.lanternpowered.server.data.key.LanternKeys
import org.lanternpowered.server.entity.LanternEntity
import org.lanternpowered.server.network.entity.parameter.ParameterList
import org.lanternpowered.server.network.value.VillagerData

class ZombieVillagerEntityProtocol<E : LanternEntity>(entity: E) : AbstractZombieEntityProtocol<E>(entity) {

    companion object {
        private val TYPE = minecraftKey("zombie_villager")
    }

    private var lastVillagerData: VillagerData? = null
    private var lastIsConverting = false

    override val mobType: NamespacedKey get() = TYPE

    override fun spawn(parameterList: ParameterList) {
        super.spawn(parameterList)
        parameterList.add(EntityParameters.ZombieVillager.VILLAGER_DATA, this.entity.villagerData)
        parameterList.add(EntityParameters.ZombieVillager.IS_CONVERTING, this.entity.get(LanternKeys.IS_CONVERTING).orElse(false))
    }

    override fun update(parameterList: ParameterList) {
        super.update(parameterList)
        val villagerData = this.entity.villagerData
        if (villagerData != this.lastVillagerData) {
            parameterList.add(EntityParameters.ZombieVillager.VILLAGER_DATA, villagerData)
            this.lastVillagerData = villagerData
        }
        val isConverting = this.entity.get(LanternKeys.IS_CONVERTING).orElse(false)
        if (isConverting != this.lastIsConverting) {
            parameterList.add(EntityParameters.ZombieVillager.IS_CONVERTING, isConverting)
            this.lastIsConverting = isConverting
        }
    }
}
